cub/device/dispatch/dispatch_histogram.cuh

```diff
@@ -346,35 +356,45 @@ public:
       }
 #endif
 
+      /**
+       * @brief Bin computation for floating point (and extended floating point) types
+       */
       template <typename T>
-      __host__ __device__ __forceinline__
-      int ComputeBin(T      sample,
-                     T      min_level,
-                     ScaleT scale,
-                     cuda::std::true_type /* is_fp */)
+      __host__ __device__ __forceinline__ int
+      ComputeBin(T sample, T min_level, ScaleT scale, ::cuda::std::true_type /* is_fp */)
       {
         return static_cast<int>((sample - min_level) * scale.reciprocal);
       }
 
-      template <typename T>
-      __host__ __device__ __forceinline__
-      int ComputeBin(T      sample,
-                     T      min_level,
-                     ScaleT scale,
-                     cuda::std::false_type /* is_fp */)
+      /**
+       * @brief Bin computation for integral types
+       */
+      template <typename T,
+                typename ::cuda::std::enable_if<::cuda::std::is_integral<T>::value, int>::type = 0>
+      __host__ __device__ __forceinline__ int
+      ComputeBin(T sample, T min_level, ScaleT scale, ::cuda::std::false_type /* is_fp */)
+      {
+        return static_cast<int>((static_cast<IntArithmeticT>(sample - min_level) *
+                                 static_cast<IntArithmeticT>(scale.fraction.bins)) /
+                                static_cast<IntArithmeticT>(scale.fraction.range));
+      }
+
+      /**
+       * @brief Bin computation for custom types
+       */
+      template <typename T,
+                typename ::cuda::std::enable_if<!::cuda::std::is_integral<T>::value, int>::type = 0>
```

Comment:
`is_integral` already means not fp. The constrain should be on the root function:
```cpp
      template <typename T>
      __host__ __device__ __forceinline__ int
      ComputeBin(T sample, T min_level, ScaleT scale, ::cuda::std::true_type /* is_fp */)
      {
        return static_cast<int>((sample - min_level) * scale.reciprocal);
      }

      /**
       * @brief Bin computation for custom types
       */
      template <typename T>
      __host__ __device__ __forceinline__ int
      ComputeBin(T sample, T min_level, ScaleT scale, ::cuda::std::false_type /* is_fp */)
      {
        return static_cast<int>(((sample - min_level) * scale.fraction.bins) /
                                scale.fraction.range);
      }

      template <typename T, typename ::cuda::std::enable_if<!::cuda::std::is_integral<T>::value, int>::type = 0>
      __host__ __device__ __forceinline__ int ComputeBin(T sample, T min_level, ScaleT scale)
      {
        return this->ComputeBin(sample, min_level, scale, ::cuda::std::is_floating_point<T>{});
      }

      template <typename T, typename ::cuda::std::enable_if<::cuda::std::is_integral<T>::value, int>::type = 0>
      __host__ __device__ __forceinline__ int ComputeBin(T sample, T min_level, ScaleT scale)
      {
        return static_cast<int>((static_cast<IntArithmeticT>(sample - min_level) *
                                 static_cast<IntArithmeticT>(scale.fraction.bins)) /
                                static_cast<IntArithmeticT>(scale.fraction.range));
      }
```

Reply:
Looks a bit cleaner, thanks. Adopted your suggestion 👍